projects/evm-compatible-bridge/federator.md

```diff
@@ -0,0 +1,245 @@
+- There should only be 1 token being sent to the MultiSig wallet.
+
+To check this format we will call the headless api to check the transaction data and validate based on the response.
+
+Since this transaction was made from Hathor it will be saved as a Hathor request on the database with the following fields:
+
+- Hathor tx_id
+  - With the tx_id, so the federators can inspect the contents.
+- Hathor token uid
+  - The equivalent EVM token can be found on the smart contract storage (it will be created if this is the first time we see this token)
+- Amount of tokens being sent to the MultiSig wallet
+- destination address (on EVM chain)
+- Height
+  - Height of the first block that confirms the transaction.
+  - Transactions in Hathor do not have a height but we can use the `first_block_height` metadata.
+
+With this we can safely wait for 5 blocks on Hathor to have passed so we can create the proposal.
+There is an event `best-block-update` which tells the current height of the network.
+This proposal will be available for the polling API as soon as it is created.
+
+### API
+
+#### GET /events
+
+This is the polling api, an event is a request to collect signatures, this means there are events to collect signatures in the MultiSig wallet and events to collect signatures on the federation contract.
+
+Both kinds of events will be returned by this endpoint, the federator will be responsible to make the appropriate calls to the coordinator service or the federation contract.
+
+The response will contain a list of events and pagination data.
+
+```ts
+type Event = {
+  eventOrigin: 'htr' | 'evm' | 'admin',
+  type: 'mint' | 'melt' | 'send', // This indicates the action on the Hathor MultiSig wallet
+  proposalId: number,
+  txHex: string,
+  // The event will have 1 of the following fields
+  evmRequestId?: number,
+  htrRequestId?: number,
+  adminRequestId?: number,
+};
+
+type ApiResponse = {
+  events: Event[],
+  hasMore: boolean,
+};
+```
+
+#### POST /sign/{proposal_id}
+
+The proposal id is referring a Hathor tx proposal where the federator has signed and this api will be used to collect the signatures.
+
+Once enough signatures are collected, the transaction will be sent to the network.
+
+The endpoint will be expecting the fields:
+
+```ts
+type RequestBody = {
+  signature: string,
+  publicKey: string,
+};
+```
+
+The public key is required so we can count how many distinct signatures were collected.
+
+#### POST /request/evm
+
+The federators will call this endpoint when an event is received in the EVM loop.
+
+The endpoint will be expecting the fields:
+
+```ts
+type RequestBody = {
+  blockHash: string,
+  txHash: string,
+  evmTokenAddr: address,
+  amount: string,
+  destinationAddress: string,
+};
+```
+
+#### POST /fulfill/htr/{htr_request_id}
+
+When a request is fulfilled, the federator will call this endpoint to inform the API that the request was fulfilled.
```

Comment:
What happens if the federator lies? Should the coordinator check on the EVM?

Or if this API is exposed publicly and someone posts the wrong info? We should always expect that possibility.

Reply:
I agree the federators can inform, but should the coordinator double check it?

Reply:
Definetely, I think any information that does not come directly from the network should be validated and this is a perfect instance of this.
The coordinator will receive multiple calls to inform that the request was fulfilled, but when the first comes through we should read from the EVM chain, check that it matches the information received and save on the database, the subsequent ones will be ignored.

Reply:
Agreed. This should be clearly stated in the design, in all APIs which need to do this check against the smart contract.

Reply:
@r4mmer was it done?